src/dawn/SIR/SIR.h

```diff
@@ -371,6 +368,12 @@ struct Value : NonCopyable {
   }
 
 private:
+  template <class T>
+  void setValue(const T& value) {
+    valueImpl_ = make_unique<ValueImpl<T>>(value);
+    type_ = TypeInfo<T>::Type;
```

Comment:
you can also completely remove `type_`: Add a member to `ValueImpl` and its base (getType) which returns `TypeInfo<T>::Type` in the implementation.